scripts/tests/test_zkai_attention_kv_quantized_softmax_receipt_gate.py

```
@@ -0,0 +1,95 @@
+        metrics = self.result["kernel_contract"]["kernel_metrics"]
+        self.assertEqual(metrics["score_rows"], 52)
+        self.assertEqual(metrics["steps"], 8)
+        self.assertEqual(len(metrics["per_step_denominators"]), 8)
+        self.assertTrue(all(denominator > 0 for denominator in metrics["per_step_denominators"]))
+        self.assertLess(metrics["max_observed_division_error_decimal"], 1.0)
+
+    def test_all_declared_mutations_reject(self):
+        base = copy.deepcopy(self.result)
+        base["mutation_results"] = []
+        for name, receipt, source, envelope, run_native in gate.mutation_cases(base, self.source, self.envelope):
+            with self.assertRaises(gate.QuantizedSoftmaxReceiptGateError, msg=name):
```

Ruff (0.15.12)

[warning] Use `pytest.raises` instead of unittest-style `assertRaises`

(PT027)